### PR TITLE
SYS-1783: Add script to extract inventory numbers from FTVA spreadsheet

### PR DESCRIPTION
**Pull request by henry-r18**
Implements [SYS-1783](https://uclalibrary.atlassian.net/browse/SYS-1783)

**Summary**
- Loads data from FTVA digital lab spreadsheet
- Matches values against inventory number pattern described by FTVA staff
  - Only outputs first match according to specified column priority
  - If multiple matches in a single input string, outputs pipe-delimited list
- Writes new Excel spreadsheet with extracted inventory numbers in the `Inventory #` column

**Testing**
I used [regex101](https://regex101.com/) to assemble and test the regular expression I used. I also wrote a short test for the regular expression using `pytest` locally, but since `pytest` isn't an existing requirement for the dev container, I didn't bother including that test in this PR.

**Running script**
From within the dev container, run: `python extract_inventory_numbers.py --data_file ${PATH_TO_DL_SPREADSHEET}`

**Note**
You may need to install the `openpyxl` package, as `pandas` uses it for reading and writing to Excel formats.

[SYS-1783]: https://uclalibrary.atlassian.net/browse/SYS-1783?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: extract_inventory_numbers.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
+import argparse
+import re
+import pandas as pd
```
Comment (akohler) [inline extract_inventory_numbers.py:3]: Since `openpxyl` is a dependency now, please add it to `requirements.txt`.  Since `pandas` is such a big package, it doesn't automatically get installed with everything, and even installing `pandas[excel]` is overkill.  Check out the [detailed documentation](https://pandas.pydata.org/pandas-docs/stable/getting_started/install.html#excel-files) for more.

For this project, looks like I added `pandas` just for Salik's `excel2xml.py`, one of several initial scripts he provided but which we haven't really used, so we didn't catch the Excel dependency issue until now. I generally work with just CSV/TSV files, but exporting the whole Google document as Excel is convenient, so it's good to bring in this dependency now.

File: extract_inventory_numbers.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+from pathlib import Path
+
+# Regular expression to match inventory numbers,
+# per FTVA specification
+INVENTORY_NUMBER_PATTERN = re.compile(r'(?:M|T|DVD|HFA|VA|XFE|XFF|XVE)\d+(?:[A-Z](?![A-Za-z]))?')
```
Comment (akohler) [inline extract_inventory_numbers.py:8]: This isn't used globally by the script, so doesn't need to be a global constant.  It can just be declared in `_extract_inventory_numbers()` where it's used.

Comment (akohler) [inline extract_inventory_numbers.py:8]: Also, please add a comment documenting the regular expression in general terms.

**Review by akohler — CHANGES_REQUESTED** [review]
@henry-r18 This looks great. The flow and the code are clear, other than the regex which should be moved to the relevant method and explained via comments.

Please do add tests.  These should be written without requiring `pytest` as a dependency, though I'm happy to hear arguments why `pytest` improves things enough over the built-in testing framework to be worth considering :)

Tests should focus on proving the regex correctly handles all good cases, and correctly handles all bad cases the specs describe and any others you can anticipate.

Thanks --Andy

File: extract_inventory_numbers.py (hunk 1 above)
Comment (henry-r18) [reply extract_inventory_numbers.py:3]: Added `openpyxl` to `requirements.txt`

File: extract_inventory_numbers.py (hunk 2 above)
Comment (henry-r18) [reply extract_inventory_numbers.py]: I moved the regex pattern and compile to its own method, to address this comment and to make it easier to import into the test module.

File: extract_inventory_numbers.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+            if not isinstance(row[column], str):
+                continue
+
+            matches = re.findall(inventory_number_pattern, row[column])
+            if matches:
+                # per FTVA spec, provide pipe-delimited string
+                # if multiple matches in a single input value
+                row['Inventory #'] = '|'.join(matches)
+                break  # we only want the first match in a row
+    print(df.iloc[existing_inventory_number_values]['Inventory #'])
```
Comment (akohler) [inline extract_inventory_numbers.py:61]: Please print a message before this, explaining what the output is.

File: tests/test_extract_inventory_numbers.py [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+from extract_inventory_numbers import _compile_regex
+
+
+class TestRegEx(unittest.TestCase):
+
+    def test_regex(self):
+        # test cases comprised of tuples
+        # where the first item is the input string
+        # and the second item is the expected output derived from FTVA specs
+        test_cases = (
```
Comment (akohler) [inline tests/test_extract_inventory_numbers.py:12]: Please add test cases for `T` and `DVD`.  Yes, they should work the same as all the others prefixes, but let the tests prove that.

File: tests/test_extract_inventory_numbers.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+            ("XFE1915MX", "XFE1915"),  # invalid suffix case
+            ("Randy_Requiem1", ""),
+            (
+                "XFE4098M_XFF104M_DamagedLives_Finals",
+                "XFE4098M|XFF104M"
+            ),  # multi-match to pipe-delimited string case
+            (
+                "XVE779T_ZVE780T_OneNightStand_WorldOfLennyBruce_CaptureFiles_SD_2997FPS_YUV",
+                "XVE779T"
+            )  # close-but-no-cigar case (ZVE780T does not have valid prefix)
```
Comment (akohler) [inline tests/test_extract_inventory_numbers.py:25]: Please tweak this comment to make it clearer what's being tested: 1 good prefix, 1 bad.  I had to read it twice to be sure it wasn't a `XVE` vs `ZVE` typo.

File: tests/test_extract_inventory_numbers.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    def test_regex(self):
+        # test cases comprised of tuples
+        # where the first item is the input string
+        # and the second item is the expected output derived from FTVA specs
+        test_cases = (
+            ("HFA27M_Reel", "HFA27M"),
+            ("VA13161T_KTLA", "VA13161T"),
+            ("M190816Medea2", "M190816"),
+            ("XFE1915MX", "XFE1915"),  # invalid suffix case
+            ("Randy_Requiem1", ""),
```
Comment (akohler) [inline tests/test_extract_inventory_numbers.py:17]: Add a comment to document that this is an invalid prefix.

File: tests/test_extract_inventory_numbers.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+            (
+                "XFE4098M_XFF104M_DamagedLives_Finals",
+                "XFE4098M|XFF104M"
+            ),  # multi-match to pipe-delimited string case
+            (
+                "XVE779T_ZVE780T_OneNightStand_WorldOfLennyBruce_CaptureFiles_SD_2997FPS_YUV",
+                "XVE779T"
+            )  # close-but-no-cigar case (ZVE780T does not have valid prefix)
+        )
```
Comment (akohler) [inline tests/test_extract_inventory_numbers.py:26]: Please add test case(s) where there's a possible valid prefix, but no valid inventory number.  For example:
* row 16360: `Max_From_DVD_H264`
* row 1227: `HARVEST3000AUDIOMAG` (I think this is invalid, but might match specs... not sure)
* row 1757: `M46293SanFernandoCLEANEXPORT3` (the `T3` might get included, though not valid - maybe another false positive)

If those last 2 are false positives, ask Thelma if there's any practical way to improve the specs.  For example, minimum number of digits?

**Review by akohler — CHANGES_REQUESTED** [review]
@henry-r18 This is almost ready.  A couple of minor comment-related tweaks, and a few more test cases, especially looking for edge cases, then I'll merge this.
Thanks --Andy

File: tests/test_extract_inventory_numbers.py (hunk 7 above)
Comment (henry-r18) [reply tests/test_extract_inventory_numbers.py:26]: @akohler, those last two cases (`HARVEST3000AUDIOMAG` and `M46293SanFernandoCLEANEXPORT3`) did indeed break my RegEx, as `T3000` substring matches the pattern in the first case and `T3` in the latter. Thanks for finding those!

I'll ask Thelma if there's a way to restrict the pattern further.

File: extract_inventory_numbers.py
Comment (henry-r18) [inline extract_inventory_numbers.py:1]: The additional regex clause, which is the first in the list of regex components, should address the edge case where capitalized words with ending letters that happen to be part of the prefix list and are followed by digits get matched by the pattern.

File: extract_inventory_numbers.py [hunk 8]
```diff
@@ -20,46 +20,28 @@ def _compile_regex() -> re.Pattern:
...
         r'(?:[A-Z](?![A-Za-z]))?'  # optional suffix 1 capital letter not followed by another letter
     ]
 
     return re.compile(''.join(regex_components))
 
 
-def _extract_inventory_numbers(df: pd.DataFrame) -> pd.DataFrame:
+def _extract_inventory_numbers(
+        value: str,
+        inventory_number_pattern: re.Pattern = _compile_regex()
```
Comment (henry-r18) [inline extract_inventory_numbers.py:34]: I moved the pattern to be a default parameter for the function, to make the function more adaptable.

File: extract_inventory_numbers.py [hunk 9]
```diff
@@ -20,46 +20,28 @@ def _compile_regex() -> re.Pattern:
...
+        r'(?:M|T|DVD|FE|HFA|VA|XFE|XFF|XVE)',  # non-capturing group of 8 prefixes defined by FTVA
         r'\d+',  # 1 or more digits, as many as possible
         r'(?:[A-Z](?![A-Za-z]))?'  # optional suffix 1 capital letter not followed by another letter
     ]
 
     return re.compile(''.join(regex_components))
 
 
-def _extract_inventory_numbers(df: pd.DataFrame) -> pd.DataFrame:
+def _extract_inventory_numbers(
```
Comment (henry-r18) [inline extract_inventory_numbers.py:32]: I revised this function quite a bit, to better suit the new extract logic.

File: tests/test_extract_inventory_numbers.py
Comment (henry-r18) [inline tests/test_extract_inventory_numbers.py:1]: I improved the inline comments on the test cases, as well as adding the cases you identified @akohler

File: extract_inventory_numbers.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+    )
+    args = parser.parse_args()
+    return args
+
+
+def _compile_regex() -> re.Pattern:
+    # Return a compiled RegEx pattern for matching inventory numbers
+    # the RegEx is comprised of three main parts, which are explained inline below
+    regex_components = [
+        r'(?<![A-Z])',  # pattern not preceded by capital letter, to mitigate false positives
```
Comment (henry-r18) [inline extract_inventory_numbers.py:23]: This additional regex clause addresses the edge case where capitalized words with ending letters that happen to be part of the prefix list and are followed by digits get matched by the pattern.

**Comment by henry-r18** [comment]
@akohler 

I confirmed with Thelma on Slack that valid inventory numbers should include 2 or more digits, so the `T1...T4` cases you identified should not match. I'm not sure how to create a regex clause that ignores the last case `T01` without explicitly excluding that specific substring. Since it appears to be the only case in the current data, I did not adjust the regex to account for it, but rather made a note that it's a known false positive.

**Latest commit**

I pushed another commit that includes the relevant change to the regex pattern, as well as:

- fixes an issue where the matches were returning out of their original order, because I used `set()` to drop duplicates
- fixes an issue where the `_extract_inventory_numbers` function returned `None` instead of an empty string
- adds tests for the new edge cases discussed above
- revised the test to target `_extract_inventory_numbers`, rather than _compile_regex, now that `_extract_inventory_numbers` is more modular

**Rebase from `main`**

I used `git pull --rebase` on my local feature branch to pull in Zoe's recent commit from `main`, but I'm not sure I did it right. Perhaps we can look at it together at stand up.

Thanks!
Henry

**Comment by henry-r18** [comment]
@akohler 

It looks to me like the regex is working as expected on `AAC442\Title_T01ASYNC_Surround`, because the last clause is meant to capture the optional suffix, which this input doesn't have. If the input was `T01A_SYNC`, the pattern would match the optional suffix, which has to be a single capital letter not followed by any other letter. But in the case at issue, we'd still want to capture T01, as it has a valid prefix and a valid number of digits, just no optional suffix. It just so happens that this inventory number isn't real, because Thelma says so.

My solution was to hard-code a list of "known false positives" (i.e. things that look like inventory numbers but aren't actually), then remove them from the matched list before returning it, as you suggested. I updated the relevant test case accordingly, and all the tests pass.

I also added to the README, which then resulted in a merge conflict. I attempted to resolve that merge conflict in yet another commit. I think all is well at this point...

-Henry

PS Thanks for walking me through git rebasing earlier—things are much clearer for me now!

**Review by akohler — APPROVED** [review]
@henry-r18 This all looks good to me now.  Your "known false positives" solution is fine, given the limited number (only 1 so far) and the messiness of the source data.  I re-checked the output and confirmed `T01` is no longer included.  All tests pass, test cases are meaningful and look complete to me.

Merging now, thanks.  @ztucker4 FYI.
